### PR TITLE
Towards autoconf #1

### PR DESCRIPTION
**Pull request by shindere**
Initially, I intended to submit the switch to autoconf as only one GPR.
However, it turns out that the resulting PR would be quite big and
not very easy to review. Moreover, since this development changes a lot
of files, I find myself spending a lot of time just rebasing my current
work as other GPRs get merged, which makes me quite slow andis, in addition,
very error-prone.

So in an attempt to simplify everybody's life, here is a first bunch of
pre-autoconf changes. This GPR is probably easier to review in a
commit-by-commit way. It contains the 4 following commits:

 1. Minor config.h cleanup 

 2. Introduction of a new C preprocessor macro, `HAS_ARCH_CODE32`. This is
 to prepare `m.h` and `s.h` to be generated from `.in` files.
 Indeed, `autoconf` prescribes that template header files must not
 contain `#ifdef` directives, so this commits avoids this by moving the
 conditional directive to `config.h`.

 3. Move `config/Makefile` to `Makefile.config`, so that the `config`
 directory can ultimately be deleted. This has already been discussed on
 caml-devel so I hope it will be okay.

 4. Use `ROOTDIR` more consistently accross Makefiles. This has led to a few
 minor enhancements, in particular in the debugger's `Makefile`.

Cc @dra27 @xavierleroy @damiendoligez

### CODE REVIEW

File: debugger/Makefile [hunk 1]
```diff
@@ -13,87 +13,66 @@
...
-	program_management.cmo \
-	frames.cmo \
-	eval.cmo \
-	show_information.cmo \
-	loadprinter.cmo \
-	parser.cmo \
-	command_line.cmo \
-	main.cmo
+DIRECTORIES=$(UNIXDIR) $(addprefix $(ROOTDIR)/,\
+  utils parsing typing bytecomp toplevel driver)
```
Comment (damiendoligez) [inline debugger/Makefile:35]: Note: you have changed the order: `$(UNIXDIR)` was at the end of the list, now it's at the front. I don't think it matters, but you might want to restore the original order just to be sure.

File: runtime/Makefile [hunk 2]
```diff
@@ -224,7 +226,7 @@ distclean: clean
 
 # Generated non-object files
 
-ld.conf: ../config/Makefile
+ld.conf: $(ROOTDIR)//Makefile.config
```
Comment (damiendoligez) [inline runtime/Makefile]: There is a spurious double `/` here.

File: runtime/Makefile [hunk 3]
```diff
@@ -277,8 +279,8 @@ caml/jumptbl.h : caml/instruct.h
 	sed -n -e '/^  /s/ \([A-Z]\)/ \&\&lbl_\1/gp' \
 	       -e '/^}/q' > $@
 
-caml/version.h : ../VERSION ../tools/make-version-header.sh
-	../tools/make-version-header.sh ../VERSION > caml/version.h
+caml/version.h : $(ROOTDIR)//tools/make-version-header.sh $(ROOTDIR)/VERSION
```
Comment (damiendoligez) [inline runtime/Makefile]: double `/` here too

File: otherlibs/dynlink/Makefile [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 # FIXME reduce redundancy by including ../Makefile
 
-include ../../config/Makefile
+include ../../Makefile.config
```
Comment (damiendoligez) [inline otherlibs/dynlink/Makefile]: Why not do the ROOTDIR transformation in this file too?

**Review by damiendoligez — APPROVED** [review]
A few small remarks but otherwise good to go.

**Comment by dra27** [comment]
Can I strongly urge that the renaming of `config/Makefile` to `Makefile.config` goes in after 4.08 has been branched? Or, alternatively, that we merge it now but revert that commit on 4.08 immediately before beta1 is cut?

**Comment by shindere** [comment]
The PR has been rebased on latest trunk and the comments have been taken
into acocunt, except for the order of directories which has been left as
originally submitted.

Merging.

**Comment by shindere** [comment]
Oooh, sorry @dra27!

Didn't you say on caml-devel that if 4.08 uses autoconf then you'll be
fine with the renaming?

**Comment by dra27** [comment]
@damiendoligez looks to have gone through the good tidy-ups in detail, so I haven't repeated that - this is all looks great to me, just with the caveat already mentioned that I'd prefer not to have yet another set of Windows instructions for 4.08 - unless the plan is for autoconf in 4.08?!

**Comment by dra27** [comment]
@shindere - crossed messages! If the 4.08 branching is holding off until all the autoconf stuff is in, then great!

**Comment by shindere** [comment]
David Allsopp (2018/09/17 06:07 -0700):
> @damiendoligez looks to have gone through the good tidy-ups in detail,
> so I haven't repeated that - this is all looks great to me, just with
> the caveat already mentioned that I'd prefer not to have yet another
> set of Windows instructions for 4.08 - unless the plan is for autoconf
> in 4.08?!

Yeah that's the plan, except that, at this stage, I am not sure the
configure script provided in 4.08 will be general enough to handle the
Windows case :( which I guess is equivalent to say there won't be
autoconf support in 4.08, seen from a Windows perspective.

Basically an issue I hadn't forseen is that autoconf in itself has no
support for finding out how to compile libraries. To achieve this, one
has to also make use of libtool and it's not obvious to me yet how easy
it will be to make libtool work harmoniously with flexlink.

So basically the idea is that we get autoconf at least for Linux in 4.08
and then for Windows either in 4.08 too if we manage to do so, or in
4.09.

**Comment by shindere** [comment]
David Allsopp (2018/09/17 06:08 -0700):
> @shindere - crossed messages! If the 4.08 branching is holding off
> until all the autoconf stuff is in, then great!

That would be ideal, I agree. Well I guess it also depends on how
promptly things go forward. I mean that if, at the time of branching, it
seems there's not much missing to reach that point, then the release can
probably be delayed a bit. If, on the countrary, a lot seems to be
missing at the time of branching then I don't think the release will be
delayed. So let's see how fast I/we can go forward. Help will definitely
be appreciated to review the related PRs given that, as mentionned in
this one, I'll try to make several smaller ones to make reviewing easier
and also to make sure when developping that the previous changes are
accepted.

**Pull request merged**